tensorflow_datasets/core/lazy_imports_lib.py

```diff
@@ -20,21 +20,93 @@
 from __future__ import division
 from __future__ import print_function
 
+from builtins import __import__ as original_import_fun
+import builtins
+import contextlib
 import importlib
+import types
 
 from tensorflow_datasets.core.utils import py_utils as utils
 
+_ERR_MSG = ("Failed importing {name}. This likely means that the dataset "
+            "requires additional dependencies that have to be "
+            "manually installed (usually with `pip install {name}`). See "
+            "setup.py extras_require.")
+
+# `lazy_imports` currently only allow the following external dependencies to
+# be used during dataset generation.
+_ALLOWED_LAZY_DEPS = [
+    "apache_beam",
+    "crepe",
+    "cv2",
+    "h5py",
+    "langdetect",
+    "librosa",
+    "matplotlib",
+    "matplotlib.pyplot",
+    "mwparserfromhell",
+    "nltk",
+    "pandas",
+    "PIL.Image",
+    "PIL.TiffImagePlugin",
+    "pretty_midi",
+    "pydub",
+    "scipy",
+    "scipy.io",
+    "scipy.ndimage",
+    "skimage",
+    "skimage.color",
+    "skimage.filters",
+    "skimage.external.tifffile",
+    "tensorflow_io",
+    "tldextract",
+]
+
 
 def _try_import(module_name):
   """Try importing a module, with an informative error message on failure."""
   try:
     mod = importlib.import_module(module_name)
     return mod
   except ImportError:
-    err_msg = ("Failed importing {name}. This likely means that the dataset "
-               "requires additional dependencies that have to be "
-               "manually installed (usually with `pip install {name}`). See "
-               "setup.py extras_require.").format(name=module_name)
+    err_msg = _ERR_MSG.format(name=module_name)
+    utils.reraise(suffix=err_msg)
+
+
+class FakeModule(types.ModuleType):
+  """A fake module which raise ImportError whenever an unknown attribute is accessed"""
+  def __init__(self, name):
+    super(FakeModule, self).__init__(name)
+
+  def __getattr__(self, attr):
+    err_msg = _ERR_MSG.format(name=self.__name__)
+    raise ImportError(err_msg)
+
+
+def _custom_import(name, *args, **kwargs):
+  try:
+    return original_import_fun(name, *args, **kwargs)
```

Comment:
original_import_fun should be the `__import__` function at the time the `with try_import` is called. Not as import time.
Otherwise it might creates issue if other libraries also try to patch `__import__`:

```
with user_custom_import():
  with tfds.core.try_import():
    import x
  import y  # Oups, use the built-in import instead of the user_custom_import
```